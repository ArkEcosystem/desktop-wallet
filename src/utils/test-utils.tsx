import { render } from "@testing-library/react";
import { i18n } from "app/i18n";
import React from "react";
import { I18nextProvider } from "react-i18next";

const WithProviders = ({ children }: { children: React.ReactNode }) => {
	return <I18nextProvider i18n={i18n}>{children}</I18nextProvider>;
};

const customRender = (component: React.ReactElement, options: any) =>
	render(component, { wrapper: WithProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
