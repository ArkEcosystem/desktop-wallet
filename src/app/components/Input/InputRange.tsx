import { useFormField } from "app/components/Form/useFormField";
import { Range } from "app/components/Range";
import { useCurrencyDisplay } from "app/hooks";
import cn from "classnames";
import React, { useEffect, useMemo } from "react";
import { getTrackBackground } from "react-range";

import { InputCurrency } from "./InputCurrency";
import { sanitizeStep } from "./utils";

type CurrencyInput = {
	display: string;
	value: string;
};

type Props = {
	disabled?: boolean;
	value?: CurrencyInput;
	avg: any;
	min: string;
	max: string;
	step: number;
	name?: string;
	magnitude?: number;
	onChange?: any;
};

export const InputRange = React.forwardRef<HTMLInputElement, Props>(
	({ min, max, step, avg, magnitude, onChange, value, disabled }: Props, ref) => {
		const { formatRange, convertToCurrency } = useCurrencyDisplay();
		const fieldContext = useFormField();
		const [inputValue, setInputValue] = React.useState<CurrencyInput>(convertToCurrency(avg));

		const rangeValues = useMemo(() => formatRange(inputValue, max), [formatRange, max, inputValue]);
		const trackBackgroundMinValue = Number(inputValue.display);
		const minValue = Math.min(Number(min), trackBackgroundMinValue);

		useEffect(() => {
			setInputValue(convertToCurrency(value));
		}, [value, convertToCurrency]);

		const handleInput = (currency: CurrencyInput) => {
			setInputValue(currency);
			onChange?.(currency);
		};

		const handleRange = (values: number[]) => {
			const amount = convertToCurrency(values[0].toString());
			onChange?.(amount);
		};

		const backgroundColor = !fieldContext?.isInvalid
			? "rgba(var(--theme-color-primary-rgb), 0.1)"
			: "rgba(var(--theme-color-danger-rgb), 0.1)";

		return (
			<InputCurrency
				disabled={disabled}
				style={
					disabled
						? undefined
						: {
								background: getTrackBackground({
									values: [trackBackgroundMinValue],
									colors: [backgroundColor, "transparent"],
									min: minValue,
									max: Number(max),
								}),
						  }
				}
				className={cn({ "pr-12": fieldContext?.isInvalid })}
				magnitude={magnitude}
				value={inputValue.display}
				ref={ref}
				onChange={handleInput}
			>
				{!disabled && Number(min) < Number(max) && (
					<div className="absolute bottom-0 px-1 w-full">
						<Range
							step={sanitizeStep({ min: Number(minValue), max: Number(max), step })}
							isInvalid={fieldContext?.isInvalid}
							min={minValue}
							max={Number(max)}
							onChange={handleRange}
							values={rangeValues}
						/>
					</div>
				)}
			</InputCurrency>
		);
	},
);

InputRange.displayName = "InputRange";
InputRange.defaultProps = {
	magnitude: 8,
};
