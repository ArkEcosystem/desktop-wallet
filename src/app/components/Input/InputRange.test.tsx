import React from "react";
import { fireEvent, render } from "test-utils";

import { InputRange } from "./InputRange";

describe("InputRange", () => {
	it("should render with default value", () => {
		const { asFragment, getByTestId } = render(<InputRange defaultValue={5} min={1} max={10} step={1} />);
		expect(getByTestId("InputCurrency")).toHaveValue("5");
		expect(getByTestId("Range__thumb")).toHaveAttribute("aria-valuenow", "5");
		expect(asFragment()).toMatchSnapshot();
	});

	it("should update the range when changing the input", () => {
		const { getByTestId } = render(<InputRange defaultValue={5} min={1} max={10} step={1} />);
		const input = getByTestId("InputCurrency");
		fireEvent.change(input, { target: { value: "9" } });
		expect(getByTestId("Range__thumb")).toHaveAttribute("aria-valuenow", "9");
	});

	it("should update the input when changing the range", () => {
		const { getByTestId } = render(<InputRange defaultValue={5} min={1} max={10} step={1} />);
		const thumb = getByTestId("Range__thumb");
		fireEvent.keyDown(thumb, { key: "ArrowRight", code: "ArrowRight" });
		expect(getByTestId("Range__thumb")).toHaveAttribute("aria-valuenow", "6");
		expect(getByTestId("InputCurrency")).toHaveValue("6");
	});
});
