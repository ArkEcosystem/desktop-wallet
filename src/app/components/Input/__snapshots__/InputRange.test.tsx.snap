// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputRange should render with default value 1`] = `
<DocumentFragment>
  <div
    class="sc-Axmtr jzhOwR"
  >
    <input
      class="sc-AxhCb clCjoe overflow-hidden shadow-sm w-full bg-theme-background appearance-none rounded border border-theme-neutral-300 py-3 px-4 text-theme-neutral-900 transition-colors duration-200 focus:outline-none focus:border-theme-primary "
      data-testid="Input"
      type="text"
      value="5"
    />
    <div
      class="absolute bottom-0 w-full px-1"
    >
      <div
        class="flex flex-wrap justify-center"
        data-testid="Range"
      >
        <div
          class="sc-AxjAm gVIMxU"
          data-testid="Range__track"
          style="transform: scale(1); cursor: pointer;"
        >
          <div
            class="sc-AxirZ iCiXDi"
            data-testid="Range__track__filled"
          >
            <div
              aria-valuemax="10"
              aria-valuemin="1"
              aria-valuenow="5"
              class="sc-AxiKw jKNfZI"
              data-testid="Range__thumb"
              draggable="false"
              role="slider"
              style="position: absolute; z-index: 0; cursor: grab; user-select: none; transform: translate(0px, 0px);"
              tabindex="0"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
